src/relay/pass/call_graph.h

```diff
@@ -0,0 +1,509 @@
+      std::unordered_map<GlobalVar, std::unique_ptr<CallGraphEntryNode>, ObjectHash,
+                         ObjectEqual>;
+  // Create iterator alias for a CallGraph object.
+  using iterator = CallGraphMap::iterator;
+  using const_iterator = CallGraphMap::const_iterator;
+
+ public:
+  /*!
+   * \brief Construct a CallGraph from a IR module.
+   *
+   * \param module The IR module
+   */
+  explicit CallGraph(IRModule module);
+
+  /*!
+   * \brief Construct from an object pointer.
+   * \param n The object pointer.
+   */
+  explicit CallGraph(ObjectPtr<Object> n) : ObjectRef(n) {}
+
+  /*! \return The begin iterator. */
+  iterator begin() {
+    auto* n = operator->();
+    CHECK(n);
+    return n->begin();
+  }
+  /*! \return The end iterator. */
+  iterator end() {
+    auto* n = operator->();
+    CHECK(n);
+    return n->end();
+  }
+  /*! \return The begin iterator. */
+  const_iterator begin() const {
+    const auto* n = operator->();
+    CHECK(n);
+    return n->begin();
+  }
+  /*! \return The end iterator. */
+  const_iterator end() const {
+    const auto* n = operator->();
+    CHECK(n);
+    return n->end();
+  }
+
+  /*!
+   * \brief Get an element from the CallGraph using a GlobalVar.
+   *
+   * \param gv The GlobalVar used for indexing.
+   *
+   * \return The fetched element.
+   */
+  const CallGraphEntryNode* operator[](const GlobalVar& gv) const {
+    const auto* n = operator->();
+    CHECK(n);
+    return (*n)[gv];
+  }
+  /*!
+   * \brief Get an element from the CallGraph using a GlobalVar.
+   *
+   * \param gv The GlobalVar used for indexing.
+   *
+   * \return The fetched element.
+   */
+  CallGraphEntryNode* operator[](const GlobalVar& gv) {
+    auto* n = operator->();
+    CHECK(n);
+    return (*n)[gv];
+  }
+  /*!
+   * \brief Get an element from the CallGraph using the global function name.
+   *
+   * \param gvar_name The global function name used for indexing.
+   *
+   * \return The fetched element.
+   */
+  const CallGraphEntryNode* operator[](const std::string& gvar_name) const {
+    const auto* n = operator->();
+    CHECK(n);
+    return (*n)[gvar_name];
+  }
+  /*!
+   * \brief Get an element from the CallGraph using the global function name.
+   *
+   * \param gvar_name The global function name used for indexing.
+   *
+   * \return The fetched element.
+   */
+  CallGraphEntryNode* operator[](const std::string& gvar_name) {
+    auto* n = operator->();
+    CHECK(n);
+    return (*n)[gvar_name];
+  }
+
+  /*! \return mutable pointers to the node. */
+  CallGraphNode* operator->() const {
+    auto* ptr = get_mutable();
+    CHECK(ptr != nullptr);
+    return static_cast<CallGraphNode*>(ptr);
+  }
+
+ private:
+  /*! \brief Overload the << operator to print a call graph. */
+  friend std::ostream& operator<<(std::ostream& os, const CallGraph&);
+};
+
+/*!
+ * \brief A node in the call graph. It maintains the edges from a caller to
+ * all callees.
+ */
+class CallGraphEntryNode {
```

Comment:
The naming of `CallGraphEntryNode` and `CallGraphEntry` is a bit confusing as they are not in the node system...

Reply:
Thanks for pointing out. Only `CallGraphNode` is in the node system. Let's use `CallGraphEntry` instead of `CallGraphEntryNode` to reduce the confusion.